### DISCUSSION TITLE
Set Focus Programmatically

### DISCUSSION THREAD

**Discussion by luna215** [molstar/molstar#570] [Q&A] [answered]
Hi, 

How would I go about providing focus on a structure programmatically? I would like to add focus to a structure as soon as I load it. Can this be done through `this.plugin.builders.structure.representation.addRepresentation` and providing the correct parameters?

Thanks!

**Comment by luna215** [comment, marked as answer]
Never mind, figured it out.

Hopefully this piece of code will help someone:
```typescript
const data = this.plugin.managers.structure.hierarchy.current.structures[1]?.cell.obj?.data;
const ligandLoci = Structure.toStructureElementLoci(data as any)

this.plugin.managers.camera.focusLoci(ligandLoci)
this.plugin.managers.interactivity.lociHighlights.highlightOnly({ loci: ligandLoci });
this.plugin.managers.interactivity.lociSelects.select({ loci: ligandLoci });
```

**Comment by mattpangenesis** [comment]
hi! i'm trying to replicate what you're doing here -- dumb question, but where did you insert this code? many thanks.

**Comment by luna215** [comment]
Hi @mattpangenesis 

Using the similar code from the [Lighting Demo](https://github.com/molstar/molstar/blob/master/src/examples/lighting/index.ts)


```typescript
class LightingDemo {
    plugin: PluginUIContext;

    private radius = 5;
    private bias = 1.1;
    private preset: Canvas3DPreset = 'occlusion';

    async init(target: string | HTMLElement) {
        this.plugin = await createPluginUI(typeof target === 'string' ? document.getElementById(target)! : target, {
            ...DefaultPluginUISpec(),
            layout: {
                initial: {
                    isExpanded: false,
                    showControls: true
                },
            },
            components: {
                controls: { left: 'none', right: 'none', top: 'none', bottom: 'none' }
            }
        });

    }
 
   highlight() {
      const data = this.plugin.managers.structure.hierarchy.current.structures[1]?.cell.obj?.data;
      const ligandLoci = Structure.toStructureElementLoci(data as any)

      this.plugin.managers.camera.focusLoci(ligandLoci)
      this.plugin.managers.interactivity.lociHighlights.highlightOnly({ loci: ligandLoci });
      this.plugin.managers.interactivity.lociSelects.select({ loci: ligandLoci });
   }

}
```

Note: that this code always highlights the second structure in the structures array (`this.plugin.managers.structure.hierarchy.current.structures[1]?.cell.obj?.data;`). 

Let me know if this helps!

**Comment by mattpangenesis** [comment]
thanks so much! for some reason, `this.plugin.managers.structure.hierarchy` shows me the state accurately, but `this.plugin.managers.structure.hierarchy.current` always gives me an empty array (when i'm working in `app.ts` in the `viewer`).

though i don't have that issue when i just start off directly in `docking-viewer`. 

very odd! but i think i can just work with docking-viewer for now. thanks again.
